npm/src/it/common-flow/test-npm-add-user-flow.groovy

```diff
@@ -0,0 +1,136 @@
+def baseScript = new GroovyScriptEngine("$project.basedir/src/it").with
+{
+    loadScriptByName('NpmIntegrationTest.groovy')
+}
+this.metaClass.mixin baseScript
+
+println "Test test-npm-add-user-flow.groovy" + "\n\n"
+
+def npmExec
+def strongboxUsername = "admin"
+def nonStrongboxUsername = "notarealuser"
+def successMsg = "logged in as "
+
+// Determine OS and appropriate commands
+if (System.getProperty("os.name").toLowerCase().contains("windows"))
+{
+    npmExec = "cmd /c npm"
+}
+else
+{
+    npmExec = "npm"
+}
+
+def executionPath = getExecutionPathRoot(project)
+def adduserDependencyRoot = executionPath.resolve("npm-add-user-test")
+
+def strongboxUser = new Credentials(strongboxUsername,"password","a@a.com")
+def nonStrongboxUser = new Credentials(nonStrongboxUsername,"zipZapZooey","derp@derpy.com")
+
+//test that npm adduser works for strongbox user
+def proc = createNpmAdduserProcess(npmExec, adduserDependencyRoot.toString())
+assert loginToNpm(proc, strongboxUser, successMsg+strongboxUsername)
+
+proc.destroy()
+
+//test that npm adduser does not work for nonstrongbox user
+proc = createNpmAdduserProcess(npmExec, adduserDependencyRoot.toString())
+assert !loginToNpm(proc, nonStrongboxUser, successMsg+nonStrongboxUsername)
+
+proc.destroy()
+
+Process createNpmAdduserProcess(String npmExec, String directory)
+{
+    def processBuilder = new ProcessBuilder(npmExec, "adduser")
+    processBuilder.directory(new File(directory))
+    return processBuilder.start()
+}
+
+boolean loginToNpm(Process proc, Credentials creds, String returnMessage)
+{
+    def output = new StringBuilder()
+    def promptReader = new PromptReader(proc.getInputStream(), output)
+    def outputThread = new Thread(promptReader)
+    def input = new OutputStreamWriter(proc.getOutputStream())
+
+    outputThread.start()
+
+    Thread.sleep(1000)
```

Comment:
Is there a way to wait until the expected input prompt has been displayed? On some machines this might happen much quicker thant on others. One second should, in theory, be enough time for this to have been displayed, but how about on a busy build server where there are many builds straining the systems. It feels like this are will be a hotspot for future synchronization fixes.

Would it be possible to make it expect these prompts and only then send our input?

Reply:
Yea I think we can use wait() and notify()